### PR TITLE
Convert the project to typescript and improve types

### PR DESCRIPTION
**Pull request by jfromaniello**
### Description

This changes convert the project to typescript to address several issues with its current type definitions and make things easier for the user.

Example of two middlewares using expressjwt:

```typescript
import { expressjwt, ExperssJwtRequest } from "express-jwt";

app.get(
  "/protected",
  expressjwt({ secret: "shhhhhhared-secret", algorithms: ["HS256"] }),
  function (req: ExperssJwtRequest, res: express.Response) {
    if (!req.auth.admin) return res.sendStatus(401);
    res.sendStatus(200);
  }
);
```

It also simplify the interface a bit by conveying to a callback pattern of `request`, `token` and `callback` in the `isRevoked` and `secret` functions.

Removes also the feature of changing the output for the decoded token, `userProperty`, `requestProperty` and `resultProperty`. The middleware puts the decoded token always on `req.auth`.

Note: this is **major** change.
### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

**Review by shellscape — COMMENTED** [review]
Thanks very much for putting this together. I have a few comments and suggestions, but nothing that would block this from moving forward.

**Comment by jfromaniello** [comment]
@shellscape Excellent feedback, since we are breaking compatibility I changed everything to promises including the parameter functions.

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+      } catch (err) {
+        throw new UnauthorizedError('invalid_token', err);
+      }
+
+      const isRevoked = options.isRevoked && await options.isRevoked(req, decodedToken) || false;
+      if (isRevoked) {
+        throw new UnauthorizedError('revoked_token', { message: 'The token has been revoked.' });
+      }
+
+      const request = req as ExperssJwtRequest<jwt.JwtPayload | string>;
```
Comment (GfxKai) [inline src/index.ts:101]:
```suggestion
      const request = req as ExpressJwtRequest<jwt.JwtPayload | string>;
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+type Params = {
+  secret: jwt.Secret | GetVerificationKey,
+  getToken?: TokenGetter,
+  isRevoked?: IsRevoked,
+  credentialsRequired?: boolean,
+} & jwt.VerifyOptions;
+
+export { UnauthorizedError } from './errors/UnauthorizedError';
+
+export type ExperssJwtRequest<T = jwt.JwtPayload> =
```
Comment (GfxKai) [inline src/index.ts:20]:
```suggestion
export type ExpressJwtRequest<T = jwt.JwtPayload> =
```

File: README.md [hunk 3]
```diff
@@ -200,20 +200,39 @@ The default behavior is to throw an error when the token is invalid, so you can
...
+    secret: "hello world !",
+    algorithms: ["HS256"],
+    credentialsRequired: false,
+  })
+);
+```
+
+## Typescript
+
+An `ExperssJwtRequest` type is provided which extends `express.Request` with the `auth` property.
```
Comment (GfxKai) [inline README.md:223]:
```suggestion
An `ExpressJwtRequest` type is provided which extends `express.Request` with the `auth` property.
```

File: README.md [hunk 4]
```diff
@@ -15,31 +15,36 @@ $ npm install express-jwt
...
 ```javascript
-var jwt = require('express-jwt');
-
-app.get('/protected',
-  jwt({ secret: 'shhhhhhared-secret', algorithms: ['HS256'] }),
-  function(req, res) {
-    if (!req.user.admin) return res.sendStatus(401);
+var { expressjwt: jwt } = require("express-jwt");
+// or ES6
+// import { expressjwt, ExperssJwtRequest } from "express-jwt";
```
Comment (GfxKai) [inline README.md:20]:
```suggestion
// import { expressjwt, ExpressJwtRequest } from "express-jwt";
```

File: README.md [hunk 5]
```diff
@@ -200,20 +200,39 @@ The default behavior is to throw an error when the token is invalid, so you can
...
+
+An `ExperssJwtRequest` type is provided which extends `express.Request` with the `auth` property.
+
+```typescript
+import { expressjwt, ExperssJwtRequest } from "express-jwt";
+
+app.get(
+  "/protected",
+  expressjwt({ secret: "shhhhhhared-secret", algorithms: ["HS256"] }),
+  function (req: ExperssJwtRequest, res: express.Response) {
```
Comment (GfxKai) [inline README.md:231]:
```suggestion
  function (req: ExpressJwtRequest, res: express.Response) {
```

File: test/jwt.test.ts [hunk 6]
```diff
@@ -0,0 +1,345 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+import * as jwt from 'jsonwebtoken';
+import * as express from 'express';
+import { expressjwt, UnauthorizedError, ExperssJwtRequest, GetVerificationKey } from '../src';
```
Comment (GfxKai) [inline test/jwt.test.ts:4]:
```suggestion
import { expressjwt, UnauthorizedError, ExpressJwtRequest, GetVerificationKey } from '../src';
```

File: test/jwt.test.ts [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+});
+
+describe('work tests', function () {
+  // var req = {} as express.Request;
+  // var res = {} as express.Response;
+
+  it('should work if authorization header is valid jwt', function (done) {
+    const secret = 'shhhhhh';
+    const token = jwt.sign({ foo: 'bar' }, secret);
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/jwt.test.ts:244]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/jwt.test.ts [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    expressjwt({ secret: secret, algorithms: ['HS256'] })(req, res, function () {
+      assert.equal(req.auth.foo, 'bar');
+      done();
+    });
+  });
+
+  it('should work if authorization header is valid with a buffer secret', function (done) {
+    const secret = Buffer.from('AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA', 'base64');
+    const token = jwt.sign({ foo: 'bar' }, secret);
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/jwt.test.ts:257]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/jwt.test.ts [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+    expressjwt({ secret: 'secretB', algorithms: ['HS256'] })(req, res, function (err) {
+      const index = err.stack.indexOf('UnauthorizedError: invalid signature')
+      assert.equal(index, 0, "Stack trace didn't include 'invalid signature' message.")
+      done();
+    });
+
+  });
+
+  it('should work with a custom getToken function', function (done) {
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/jwt.test.ts:302]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/jwt.test.ts [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+      algorithms: ['HS256'],
+      getToken: getTokenFromQuery
+    })(req, res, function () {
+      assert.equal(req.auth.foo, 'bar');
+      done();
+    });
+  });
+
+  it('should work with a secretCallback function that accepts header argument', function (done) {
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/jwt.test.ts:326]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/multitenancy.test.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
+import * as jwt from 'jsonwebtoken';
+import * as express from 'express';
+import { expressjwt, ExperssJwtRequest, GetVerificationKey } from '../src';
```
Comment (GfxKai) [inline test/multitenancy.test.ts:3]:
```suggestion
import { expressjwt, ExpressJwtRequest, GetVerificationKey } from '../src';
```

File: test/multitenancy.test.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import assert from 'assert';
+
+describe('multitenancy', function () {
+  const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/multitenancy.test.ts:7]:
```suggestion
  const req = {} as ExpressJwtRequest;
```

File: test/revocation.test.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
+import * as jwt from 'jsonwebtoken';
+import * as express from 'express';
+import { expressjwt, ExperssJwtRequest } from '../src';
```
Comment (GfxKai) [inline test/revocation.test.ts:3]:
```suggestion
import { expressjwt, ExpressJwtRequest } from '../src';
```

File: test/revocation.test.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    algorithms: ['HS256'],
+    isRevoked: async (req, token) => {
+      const isRevoked = typeof token.payload !== 'string' &&
+        token.payload.jti === revoked_id;
+      return isRevoked;
+    }
+  });
+
+  it('should throw if token is revoked', function () {
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/revocation.test.ts:22]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/revocation.test.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+    middleware(req, res, function (err) {
+      assert.ok(err);
+      assert.equal(err.code, 'revoked_token');
+      assert.equal(err.message, 'The token has been revoked.');
+    });
+  });
+
+  it('should work if token is not revoked', function () {
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/revocation.test.ts:38]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/revocation.test.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    req.headers = {};
+    req.headers.authorization = 'Bearer ' + token;
+
+    middleware(req, res, function () {
+      assert.equal(req.auth.foo, 'bar');
+    });
+  });
+
+  it('should throw if error occurs checking if token is revoked', function (done) {
+    const req = {} as ExperssJwtRequest;
```
Comment (GfxKai) [inline test/revocation.test.ts:52]:
```suggestion
    const req = {} as ExpressJwtRequest;
```

File: test/string_token.test.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
+import * as jwt from 'jsonwebtoken';
+import * as express from 'express';
+import { expressjwt, ExperssJwtRequest } from '../src';
```
Comment (GfxKai) [inline test/string_token.test.ts:3]:
```suggestion
import { expressjwt, ExpressJwtRequest } from '../src';
```

File: test/string_token.test.ts [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+import assert from 'assert';
+
+
+describe('string tokens', function () {
+  const req = {} as ExperssJwtRequest<string>;
```
Comment (GfxKai) [inline test/string_token.test.ts:8]:
```suggestion
  const req = {} as ExpressJwtRequest<string>;
```

File: README.md [hunk 19]
```diff
@@ -200,20 +200,39 @@ The default behavior is to throw an error when the token is invalid, so you can
...
+  })
+);
+```
+
+## Typescript
+
+An `ExperssJwtRequest` type is provided which extends `express.Request` with the `auth` property.
+
+```typescript
+import { expressjwt, ExperssJwtRequest } from "express-jwt";
```
Comment (GfxKai) [inline README.md:226]:
```suggestion
import { expressjwt, ExpressJwtRequest } from "express-jwt";
```

**Review by GfxKai — CHANGES_REQUESTED** [review]
Came across this PR as I was trying to debug some typing issues—problem solved, thanks!! Just adding a review as I noticed a typo 🦞

**Comment by jfromaniello** [comment]
Good catch @GfxKai ! Thank you very much ❤️

**Comment by shellscape** [comment]
@jfromaniello I think we're good 🎉

**Comment by jfromaniello** [comment]
I just released it as `v7.0.0`.

**Pull request merged**